thunder/distributed/transforms/ddp.py

```diff
@@ -202,3 +202,66 @@ def optimize_allreduce_in_ddp_backward(
     )
     utils.check(visit_callable.has_replaced_return, lambda: "")
     return updated_bwd_trace
+
+
+def apply_bucketing_to_grad_allreduce(trace: TraceCtx) -> TraceCtx:
+    """Apply Bucketing to Gradient AllReduce.
+
+    This method takes a trace which could be one representing forward or backward of an
+    :class:`~torch.nn.Module` and also one representing a method, not a PyTorch Module.
+    Then this applies bucketing to the input when it is a trace defining a DDP backward computation
+    where :func:`~thunder.distributed.all_reduce` is applied to gradients.
+    Otherwise, this method returns the input trace as is.
+
+    Args:
+        trace: A :class:`thunder.core.trace.TraceCtx`.
+
+    Returns:
+        :class:`thunder.core.trace.TraceCtx`
+    """
+
+    from torch.distributed import ProcessGroup
+    from thunder.core.compile_data import get_compile_data
+
+    if get_skip_data_parallel_grad_sync():
+        return trace
```

Comment:
It should be possible to create and jit a program that includes allreduce in a non-data parallel setting. This pass should work if there are allreduces in the trace.
What do you think of scanning the trace probably from the end of the bound symbols list to optimize for the backward case to check whether there's an allreduce in the trace or not. Return early if there's less than one allreduce.

Reply:
> a program that includes allreduce in a non-data parallel setting.

iirc the bucketing got puzzled by such trace so that I added a check in https://github.com/Lightning-AI/lightning-thunder/commit/896b40131484de8637586bd2de9be2a55fd9496f.

> What do you think of scanning the trace probably from the end of the bound symbols list to optimize for the backward case to check whether there's an allreduce in the trace or not. Return early if there's less than one allreduce

I don't think I understand this as I thought the following part does this.

Reply:
I'm sorry what I meant is that we should make (in the future, not now) the bucketing transform applicable to any trace. As you pointed out in the other comment (https://github.com/Lightning-AI/lightning-thunder/pull/222/files#r1570536852) this pass works only if it's marked for ddp today. We should allow people to write Thunder code involving allreduce in the user script to be optimized by Thunder.